src/pam/main.cc

```diff
@@ -0,0 +1,401 @@
+    return PAM_AUTHINFO_UNAVAIL;
+  }
+
+  // Stop if we're in a remote shell and configured to exit
+  if (reader.GetBoolean("core", "ignore_ssh", true)) {
+    if (getenv("SSH_CONNECTION") != nullptr ||
+        getenv("SSH_CLIENT") != nullptr || getenv("SSHD_OPTS") != nullptr) {
+      syslog(LOG_INFO, "Skipped authentication, SSH session detected");
+      return PAM_AUTHINFO_UNAVAIL;
+    }
+  }
+
+  // Try to detect the laptop lid state and stop if it's closed
+  if (reader.GetBoolean("core", "ignore_closed_lid", true)) {
+    glob_t glob_result;
+
+    // Get any files containing lid state
+    int return_value =
+        glob("/proc/acpi/button/lid/*/state", 0, nullptr, &glob_result);
+
+    if (return_value != 0) {
+      syslog(LOG_ERR, "Failed to read files from glob: %d", return_value);
+      if (errno != 0) {
+        syslog(LOG_ERR, "Underlying error: %s (%d)", strerror(errno), errno);
+      }
+      globfree(&glob_result);
+    }
+
+    for (size_t i = 0; i < glob_result.gl_pathc; i++) {
+      std::ifstream file(std::string(glob_result.gl_pathv[i]));
+      std::string lid_state;
+      std::getline(file, lid_state, static_cast<char>(file.eof()));
+
+      if (lid_state.find("closed") != std::string::npos) {
+        globfree(&glob_result);
+
+        syslog(LOG_INFO, "Skipped authentication, closed lid detected");
+        return PAM_AUTHINFO_UNAVAIL;
+      }
+    }
+
+    globfree(&glob_result);
+  }
+
+  // If enabled, send a notice to the user that facial login is being attempted
+  if (reader.GetBoolean("core", "detection_notice", false)) {
+    if ((conv_function(PAM_TEXT_INFO,
+                       S("Attempting facial authentication").c_str())) !=
+        PAM_SUCCESS) {
+      syslog(LOG_ERR, "Failed to send detection notice");
+    }
+  }
+
+  // Get the username from PAM, needed to match correct face model
+  char *username = nullptr;
+  if ((pam_res = pam_get_user(pamh, const_cast<const char **>(&username),
+                              nullptr)) != PAM_SUCCESS) {
+    syslog(LOG_ERR, "Failed to get username");
+    return pam_res;
+  }
+
+  posix_spawn_file_actions_t file_actions;
+  posix_spawn_file_actions_init(&file_actions);
+
+  const char *const args[] = {"/usr/bin/python3", // NOLINT
+                              "/lib/security/howdy/compare.py", username,
+                              nullptr};
+  pid_t child_pid;
+
+  // Start the python subprocess
+  if (posix_spawnp(&child_pid, "/usr/bin/python3", &file_actions, nullptr,
+                   const_cast<char *const *>(args), nullptr) > 0) {
+    syslog(LOG_ERR, "Can't spawn the howdy process: %s (%d)", strerror(errno),
+           errno);
+    return PAM_SYSTEM_ERR;
+  }
+
+  // NOTE: We should replace mutex and condition_variable by atomic wait, but
+  // it's too recent (C++20)
+  std::mutex m;
+  std::condition_variable cv;
+  std::atomic<ConfirmationType> confirmation_type(ConfirmationType::Unset);
+
```

Comment:
Is it a problem if we force building against a recent compiler version? GCC and Clang (11) already have support for atomic waiting, but I'm not sure of how these versions are available.